src/Domain/Structure/Hypercube.hpp

```diff
@@ -0,0 +1,172 @@
+  explicit HypercubeElement(Indices... indices) noexcept
+      : index_{{static_cast<Side>(indices)...}} {}
+
+  template <size_t LocalElementDim = ElementDim,
+            Requires<LocalElementDim == 1> = nullptr>
+  HypercubeElement(size_t dim_in_parent,
+                   std::array<Side, HypercubeDim - 1> index) noexcept;
+
+  // @{
+  /// The parent hypercube's dimensions that this element shares
+  const std::array<size_t, ElementDim>& dimensions_in_parent() const noexcept;
+
+  template <size_t LocalElementDim = ElementDim,
+            Requires<LocalElementDim == 1> = nullptr>
+  size_t dimension_in_parent() const noexcept;
+  // @}
+
+  // @{
+  /// Whether this element is located on the lower or upper side in those
+  /// dimensions that it does not share with its parent hypercube
+  const std::array<Side, HypercubeDim - ElementDim>& index() const noexcept;
+
+  const Side& side_in_parent_dimension(size_t d) const noexcept;
+
+  template <size_t NumIndices = HypercubeDim - ElementDim,
+            Requires<NumIndices == 1> = nullptr>
+  const Side& side() const noexcept {
+    return index_[0];
+  }
+  // @}
+
+  bool operator==(const HypercubeElement& rhs) const noexcept {
+    return dimensions_in_parent_ == rhs.dimensions_in_parent_ and
+           index_ == rhs.index_;
+  }
+
+  bool operator!=(const HypercubeElement& rhs) const noexcept {
+    return not(*this == rhs);
+  }
+
+ private:
+  std::array<size_t, ElementDim> dimensions_in_parent_{};
+  std::array<Side, HypercubeDim - ElementDim> index_{};
+};
+
+template <size_t ElementDim, size_t HypercubeDim>
+std::ostream& operator<<(
+    std::ostream& os,
+    const HypercubeElement<ElementDim, HypercubeDim>& element) noexcept;
+
+/// A vertex in a `Dim`-dimensional hypercube
+template <size_t Dim>
+using Vertex = HypercubeElement<0, Dim>;
+
+/// An edge in a `Dim`-dimensional hypercube
+template <size_t Dim>
+using Edge = HypercubeElement<1, Dim>;
+
+/// A face in a `Dim`-dimensional hypercube
+template <size_t Dim>
+using Face = HypercubeElement<2, Dim>;
+
+/// A cell in a `Dim`-dimensional hypercube
+template <size_t Dim>
+using Cell = HypercubeElement<3, Dim>;
+
+/*!
+ * \brief Iterator over all `ElementDim`-dimensional elements on the boundary of
+ * a `HypercubeDim`-dimensional hypercube.
+ *
+ * \see `HypercubeElement`
+ */
+template <size_t ElementDim, size_t HypercubeDim>
+struct HypercubeElementsIterator {
+  static_assert(
+      ElementDim <= HypercubeDim,
+      "Hypercube element dimension must not exceed hypercube dimension.");
+
+ public:
+  static constexpr size_t num_indices = HypercubeDim - ElementDim;
+
+  /// The number of `ElementDim`-dimensional elements on the boundary of a
+  /// `HypercubeDim`-dimensional hypercube.
+  static constexpr size_t size() noexcept {
+    return two_to_the(num_indices) * factorial(HypercubeDim) /
+           factorial(ElementDim) / factorial(num_indices);
```

Comment:
If you intend to use this for high dimensions, this should be written in a way less susceptible to overflow.  My guess is you don't expect to use it that way, in which case this is fine.

Reply:
I'm only instantiating the template for `HypercubeDim<=3`, that's probably all we need.